libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs

```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
                     }
                     else
                     {
-                        await turnContext.SendActivityAsync(new Activity { Type = ActivityTypesEx.InvokeResponse, Value = new InvokeResponse { Status = 404 } }, cancellationToken).ConfigureAwait(false);
+                        await this.SendInvokeResponseAsync(turnContext, cancellationToken, 404).ConfigureAwait(false);
                     }
                 }
                 catch
                 {
-                    await turnContext.SendActivityAsync(new Activity { Type = ActivityTypesEx.InvokeResponse, Value = new InvokeResponse { Status = 500 } }, cancellationToken).ConfigureAwait(false);
+                    await this.SendInvokeResponseAsync(turnContext, cancellationToken, 500).ConfigureAwait(false);
+                }
+            }
+            else if (IsTokenExchangeRequestInvoke(turnContext))
+            {
+                var tokenExchangeRequest = ((JObject)turnContext.Activity.Value).ToObject<TokenExchangeInvokeRequest>();
+
+                if (tokenExchangeRequest == null)
+                {
+                    await this.SendInvokeResponseAsync(
+                        turnContext,
+                        cancellationToken,
+                        400,
+                        new TokenExchangeInvokeResponse()
+                        {
+                            Id = tokenExchangeRequest.Id,
+                            ConnectionName = _settings.ConnectionName,
+                            FailureDetail = "Missing TokenExchangeInvokeRequest",
+                        }).ConfigureAwait(false);
+                }
+                else if (tokenExchangeRequest.ConnectionName != _settings.ConnectionName)
+                {
+                    await this.SendInvokeResponseAsync(
+                        turnContext,
+                        cancellationToken,
+                        400,
+                        new TokenExchangeInvokeResponse()
+                        {
+                            Id = tokenExchangeRequest.Id,
+                            ConnectionName = _settings.ConnectionName,
+                            FailureDetail = "Invalid ConnectionName in the TokenExchangeInvokeRequest",
+                        }).ConfigureAwait(false);
+                }
+                else if (!(turnContext.Adapter is ITokenExchangeProvider adapter))
+                {
+                    await this.SendInvokeResponseAsync(
+                           turnContext,
+                           cancellationToken,
+                           502,
+                           new TokenExchangeInvokeResponse()
+                           {
+                               Id = tokenExchangeRequest.Id,
+                               ConnectionName = _settings.ConnectionName,
+                               FailureDetail = "Bot Adapter does not support token exchange.",

```

Comment:
Error message: Could recommend how to fix it? 'Bot adapter does not support token exchange. Upgrade your bot to the latest version of the Bot Framework SDK.' Or something that helps users indicating what to do?